### ISSUE TITLE
IPv6 Compliance RFC4861: Neighbor Solicitation Origination, Address Resolution [v6LC.2.1.5 Part A]

### ISSUE THREAD

**Issue by LiveFreeAndRoam** [systemd/systemd#28434]
### systemd version the issue has been seen with

250

### Used distribution

Embedded Linux - Debian-based

### Linux kernel version used

5.15.71-rt51+gc36e774d0d9a

### CPU architectures issue was seen on

aarch64

### Component

systemd-networkd

### Expected behaviour you didn't see

Verify that a node properly originates Neighbor Solicitations when trying to resolve the address of a neighbor.

RFC4861 section 7.2.2 states:

```
Retransmissions MUST be rate-limited to at most one solicitation per
neighbor every RetransTimer milliseconds.
```

### Unexpected behaviour you saw

Host continued to send NS every 1 second, even though the 2nd RA specified a retransmit interval of 5 seconds.

Here is a blow-by-blow look at the protocol exchanges...

```
$ tcpdump -i eno0 -ttttt -nn -vvvv ip6
tcpdump: listening on eno0, link-type EN10MB (Ethernet), snapshot length 262144 bytes
```

#### Router sends RA with Retrans Timer 1 second
```
 00:00:00.000000 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 56) fe80::200:10ff:fe10:1060 > ff02::1: [icmp6 sum ok] ICMP6, router advertisement, length 56
        hop limit 64, Flags [none], pref medium, router lifetime 9000s, reachable time 30000ms, retrans timer 1000ms
          prefix info option (3), length 32 (4): 2001:2:0:1000::/64, Flags [onlink, auto], valid time 65535s, pref. time 65535s
            0x0000:  40c0 0000 ffff 0000 ffff 0000 0000 2001
            0x0010:  0002 0000 1000 0000 0000 0000 0000
          mtu option (5), length 8 (1):  1500
            0x0000:  0000 0000 05dc
 00:00:00.010470 IP6 (hlim 1, next-header Options (0) payload length: 56) fe80::9640:c9ff:fed6:dcd7 > ff02::16: HBH (rtalert: 0x0000) (padn) [icmp6 sum ok] ICMP6, multicast listener report v2, 2 group record(s) [gaddr ff02::1:3 to_ex { }] [gaddr ff02::1:ffd6:dcd7 to_ex { }]
 00:00:00.838592 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 32) :: > ff02::1:ffd6:dcd7: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has 2001:2:0:1000:9640:c9ff:fed6:dcd7
          unknown option (14), length 8 (1): 
            0x0000:  6af5 a604 3a04
 00:00:00.870762 IP6 (hlim 1, next-header Options (0) payload length: 56) fe80::9640:c9ff:fed6:dcd7 > ff02::16: HBH (rtalert: 0x0000) (padn) [icmp6 sum ok] ICMP6, multicast listener report v2, 2 group record(s) [gaddr ff02::1:3 to_ex { }] [gaddr ff02::1:ffd6:dcd7 to_ex { }]
```

#### Test Node sends ping request
Host responds by sending NS every 1 second, as expected.

```
 00:00:15.128100 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 10) fe80::200:10ff:fe10:1080 > fe80::9640:c9ff:fed6:dcd7: [icmp6 sum ok] ICMP6, echo request, id 0, seq 0
 00:00:15.128659 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 32) fe80::9640:c9ff:fed6:dcd7 > ff02::1:ff10:1080: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::200:10ff:fe10:1080
          source link-address option (1), length 8 (1): 94:40:c9:d6:dc:d7
            0x0000:  9440 c9d6 dcd7
 00:00:16.134440 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 32) fe80::9640:c9ff:fed6:dcd7 > ff02::1:ff10:1080: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::200:10ff:fe10:1080
          source link-address option (1), length 8 (1): 94:40:c9:d6:dc:d7
            0x0000:  9440 c9d6 dcd7
 00:00:17.158664 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 32) fe80::9640:c9ff:fed6:dcd7 > ff02::1:ff10:1080: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::200:10ff:fe10:1080
          source link-address option (1), length 8 (1): 94:40:c9:d6:dc:d7
            0x0000:  9440 c9d6 dcd7
```

#### Router sends RA with Retrans Timer 5 seconds

```
 00:00:40.052791 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 56) fe80::200:10ff:fe10:1060 > ff02::1: [icmp6 sum ok] ICMP6, router advertisement, length 56
        hop limit 64, Flags [none], pref medium, router lifetime 9000s, reachable time 30000ms, retrans timer 5000ms
          prefix info option (3), length 32 (4): 2001:2:0:1000::/64, Flags [onlink, auto], valid time 65535s, pref. time 65535s
            0x0000:  40c0 0000 ffff 0000 ffff 0000 0000 2001
            0x0010:  0002 0000 1000 0000 0000 0000 0000
          mtu option (5), length 8 (1):  1500
            0x0000:  0000 0000 05dc
 00:00:40.062490 IP6 (hlim 1, next-header Options (0) payload length: 56) fe80::9640:c9ff:fed6:dcd7 > ff02::16: HBH (rtalert: 0x0000) (padn) [icmp6 sum ok] ICMP6, multicast listener report v2, 2 group record(s) [gaddr ff02::1:3 to_ex { }] [gaddr ff02::1:ffd6:dcd7 to_ex { }]
 00:00:40.422750 IP6 (hlim 1, next-header Options (0) payload length: 56) fe80::9640:c9ff:fed6:dcd7 > ff02::16: HBH (rtalert: 0x0000) (padn) [icmp6 sum ok] ICMP6, multicast listener report v2, 2 group record(s) [gaddr ff02::1:3 to_ex { }] [gaddr ff02::1:ffd6:dcd7 to_ex { }]
```

#### TN Sends Ping Request
Host responds by sending NS ever 1 second, instead of every 5 seconds.

```
 00:01:05.989148 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 10) fe80::200:10ff:fe10:1081 > fe80::9640:c9ff:fed6:dcd7: [icmp6 sum ok] ICMP6, echo request, id 0, seq 0
 00:01:05.989721 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 32) fe80::9640:c9ff:fed6:dcd7 > ff02::1:ff10:1081: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::200:10ff:fe10:1081
          source link-address option (1), length 8 (1): 94:40:c9:d6:dc:d7
            0x0000:  9440 c9d6 dcd7
 00:01:07.014454 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 32) fe80::9640:c9ff:fed6:dcd7 > ff02::1:ff10:1081: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::200:10ff:fe10:1081
          source link-address option (1), length 8 (1): 94:40:c9:d6:dc:d7
            0x0000:  9440 c9d6 dcd7
 00:01:08.038507 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 32) fe80::9640:c9ff:fed6:dcd7 > ff02::1:ff10:1081: [icmp6 sum ok] ICMP6, neighbor solicitation, length 32, who has fe80::200:10ff:fe10:1081
          source link-address option (1), length 8 (1): 94:40:c9:d6:dc:d7
            0x0000:  9440 c9d6 dcd7
```

#### Clean Up

```
 00:01:35.674335 IP6 (hlim 255, next-header ICMPv6 (58) payload length: 48) fe80::200:10ff:fe10:1060 > ff02::1: [icmp6 sum ok] ICMP6, router advertisement, length 48
        hop limit 64, Flags [none], pref medium, router lifetime 0s, reachable time 0ms, retrans timer 1000ms
          prefix info option (3), length 32 (4): 2001:2:0:1000::/64, Flags [onlink], valid time 0s, pref. time 0s
            0x0000:  4080 0000 0000 0000 0000 0000 0000 2001
            0x0010:  0002 0000 1000 0000 0000 0000 0000
^C
16 packets captured
16 packets received by filter
0 packets dropped by kernel
```

### Steps to reproduce the problem

This requires a protocol exchange.

#### Retransmit Timer 1 second
1. Router sends an RA with "Retransmit Interval" set to 1 second.
2. TN sends a ping request
3. Host sends NS every 1 second

#### Retransmit Timer 5 seconds (ignored by networkd)
5. Router sends an RA with "Retransmit Interval" set to 5 seconds.
6. TN sends a ping request
7. Host sends NS every 1 second, when it should send every 5 seconds.

**Comment by LiveFreeAndRoam** [comment]
I have attached PCAP files and log files for your convenience.

[2_1_5a - Neighbor Solicitation Origination, Address Resolution.zip](https://github.com/systemd/systemd/files/12074019/2_1_5a.-.Neighbor.Solicitation.Origination.Address.Resolution.zip)

**Comment by LiveFreeAndRoam** [comment]
@ssahani, @keszybz @yuwata 

#28861 does not fix this issue.

This issue is related to Neighbor Solicitation Origination, whereas #28861 addresses RA handling of "AdvRetransTimer" (it stores the RA's "AdvRetransTimer" in the Host's "RetransTimer" variable, see [RFC4861 6.3.2](https://www.rfc-editor.org/rfc/rfc4861#section-6.3.2)). I would guess #28861 was fixed as an addendum to #28449, which addressed RA handling of CurHopLimit.  

Please re-open this issue.

**Comment by LiveFreeAndRoam** [comment]
Resolved by: https://github.com/systemd/systemd/pull/30490

**Comment by yuwata** [comment]
Should be fixed by #31041.

**Comment by LiveFreeAndRoam** [comment]
> Should be fixed by #31041.

I can confirm that it is fixed.  I have rerun the IPv6 Conformance test 2.1.5a and 2.1.5b.  Both tests now pass.
